tests/lib/rules/no-this-in-sfc.js

```diff
@@ -193,15 +229,6 @@ ruleTester.run('no-this-in-sfc', rule, {
       return null;
     }`,
     errors: [{message: ERROR_MESSAGE}]
-  }, {
-    code: `
-    function Foo(props) {
-      if (this.props.foo) {
-        something();
-      }
-      return null;
-    }`,
```

Comment:
this is a valid component, and must be treated as one.

Reply:
But what exactly makes this function a component? this could perfectly pass as a function [constructor](https://css-tricks.com/understanding-javascript-constructors/).

Reply:
That's the challenge with component detection.

Legacy constructors tho are fine to be false positives - nobody should be writing those anymore, especially not in parts an app using React (and thus linted by this plugin). Constructors should be written with `class`.

The false positives that we need to focus on avoiding are render props, or normal non-component functions to return jsx for refactoring purposes inside a component.

In other words, "returns jsx in any branch" means it's definitely a component; and "returns something a component *can't* in any branch" means it's definitely not; there's ambiguity in between.

Reply:
> The false positives that we need to focus on avoiding are render props, or normal non-component functions to return jsx for refactoring purposes inside a component.

In that case I think I'll close this PR since other than the uppercase `F` there is not really a way of knowing if that function is really a component or other thing.

Reply:
If it returns a class or a function inline, it's not a component, but sure.